### ISSUE TITLE
Editing a poll's message after it has been published causes an issue with the poll results. 

### ISSUE THREAD

**Issue by jpittman** [mastodon/mastodon#24997]
### Steps to reproduce the problem

1. Run a short poll.
2. Document the results.
3. Edit the original poll's message.
4. Verify if the results of the poll are skewed in some way.

### Expected behaviour

Editing the text of the post should not have any impact on the poll's data.

### Actual behaviour

The poll is reset. Results are zeroed out and the poll is reset to it's time duration limit.

### Detailed description

Here's the issue:

Editing a poll's message after it has been published causes an issue with the poll results. _This edit is about the poll description text in the body of the message it's posted with, NOT the poll option text._

When I discovered the issue, I initially decided to check the client I was using (Ivory for IOS running a dev build) and compare it to the web interface for dice.camp (running 4.1.2). As it would turn out, both instances of the polls were reporting the same incorrect data. So not client facing. 

So I ran the following test:

 - Run a short poll.
 - Document the results after it ended.
 - Edit the original poll's message.
 - Verify if the results of the poll are skewed in some way.

This produced the following results:

 - Poll restarts with it's time limit.
 - Vote Count reset to 0.
 - Prior votes are lost.

Searched both open and closed issues here and didn't find anything that look like it directly related. I did find a Reddit post that was experiencing something similar:

https://www.reddit.com/r/Mastodon/comments/zlej1b/how_to_recalculate_poll_results/

which had me run a second test:

 - Run a short poll.
 - Document the results half way through the duration.
 - Edit the original poll's message.
 - Document results after edit.
 - Verify if the results of the poll are skewed in some way.

I experienced the same result.

I posted on the dice.camp instance of mastodon about this issue as I was working it through:

https://dice.camp/@videostorecowboy/110357950338837603

There are screenshots and links to the original posts there.

### Specifications

Server: https://dice.camp
Mastodon: 4.1.2

**Comment by stevenjlm** [comment]
I couldn't reproduce this in my dev-container instance.. The poll percentages persist so long as I don't edit the poll options. Maybe this is instance-specific, or I'm missing something?

**Comment by jpittman** [comment]
In the time since I wrote this issue up, it happened again. Screenshots attached to this message:

https://dice.camp/@videostorecowboy/110375995457043208

I can reach out directly to the server admin and see if there is something amiss. I’ll also give it a try on a different server that I have an account on and see what I can find out. 

Can we leave this issue open while I try to track down more details?

**Comment by RianQuenlin** [comment]
Are you sure this isn't intentional? Being able to edit the context of a poll while retaining the results is not a good thing.

**Comment by Ember-ruby** [comment]
> Are you sure this isn't intentional? Being able to edit the context of a poll while retaining the results is not a good thing.

definitely, though, mastodon already notifies you if a post you boost is edited, maybe that could be done if a poll is edited?

**Comment by stevenjlm** [comment]
@RianQuenlin @Ember-ruby 
These are really good points.. As far as I can tell the default behavior actually does allow you to edit the context without resetting the data (according to some tests I did in a dev container). And, if you boost a poll that subsequently gets edited, you will not receive a notification. There will just be a star (*) next to the remaining time on the poll.

**Comment by Sagelt** [comment]
Instance admin for dice.camp where the poll was posted: we're running on masto.host using their default image without any modifications. If we somehow broken this it's broken for all masto.host instances.

**Comment by jpittman** [comment]
> Are you sure this isn't intentional? Being able to edit the context of a poll while retaining the results is not a good thing.

I would disagree.  It makes the assumption that anyone posting has bad intentions just by posting a poll.

For example, if someone edits an option in the poll itself, then yes I would expect this to be the behavior. This would prevent from someone altering the results in some way based upon the options of the poll itself.

However, if someone makes a mistake in spelling of a word, the intent of the poll itself (the results) shouldn't suffer as a penalty. 

> definitely, though, mastodon already notifies you if a post you boost is edited, maybe that could be done if a poll is edited?

Would this serve as a way to prompt the user who gets the notification to resubmit their poll answer?

**Comment by stevenjlm** [comment]
Ok.. this is odd. It looks like resetting the number of voters at least is very much [intentional](https://github.com/mastodon/mastodon/pull/21700) actually.

**Comment by stevenjlm** [comment]
@jpittman @Sagelt do you think it's possible that you switched the poll from multiple select to single select? I saw in the code that the poll will definitely reset everything if you (1) change the text of the poll options or (2) click on the radio button next to the option and switch it to a checkbox. Furthermore, when you look in the edit history for a poll, the poll options _always_ show up as radio buttons, even if that was changed. This in of itself, I believe, is worth fixing (that is, fixing the edit history view).

People [already mentioned](https://github.com/mastodon/mastodon/pull/21700#issuecomment-1339137030) that we should have a warning when you make changes to a poll that will reset it. This has not been implemented, so that would be another thing to fix.

Personally, I think that for consistency, if there's a notification for a changed post, there should also be a notification for a changed poll. Whether or not it tells you to vote again depends on my last point..

Resetting the poll after changing the options I think is pretty clear, but after changing the context.. I'm not sure. Resetting seems to be the safe option, although the "pollster" will lose data. I'm still pretty new to working on this repo. I'm going to ask around and see how we can get input from maintainers or others on this.

**Comment by stevenjlm** [comment]
@ClearlyClaire @Gargron do you have some guidance on what do about the last point?

**Comment by FediVideos** [comment]
Just had this happen to me:

1. I had a poll running on the FediTips account
2. Over a thousand people had responded
3. Some people asked for clarification on the wording so I edited the poll labels
4. Without any warning, the poll instantly reset to zero responses

I can see why the reset might be desirable in order to prevent abuse of polls, but there should be some kind of warning that this will happen. It's not obvious at all (and personally I feel a right idiot for not realising this would happen 😆 ).

**Comment by nehemiagurl** [comment]
I agree with @FediVideos that a warning is needed. the people in this thread are far from being the only people this happened to.

**Comment by kjuh78** [comment]
Imagine a poll with the „yes“ and „no“ as possible answers. changing the text could easily turn the result on its head. While fixing tyos would be

**Comment by FediVideos** [comment]
> Imagine a poll with „yes“ and „no“ as possible answers. A minor change to the polls text could easily turn the results upside down. While a possibility to fix typos is convenient, any possibilty to make changes conserving the answers opens a way to major manipulation.

Totally agree, but some warning about this would be appreciated! This can totally catch people out, especially as editing the main text above the poll *doesn't* reset the poll numbers 😅

**Comment by nehemiagurl** [comment]
@kjuh78 you're not wrong, but that doesn't mean the current situation can't (quite easily) be improved upon.
either of the two options, a warning before editing like FediVideo suggested, or keeping the results but adding a notification to all responders after it was edited a la boost like Ember-ruby suggested, would be better than the current situation.

**Comment by nehemiagurl** [comment]
> especially as editing the main text above the poll doesn't reset the poll numbers

doesn't it? I was sure it did.

**Comment by FediVideos** [comment]
> > especially as editing the main text above the poll doesn't reset the poll numbers
> 
> doesn't it? I was sure it did.

Nope, it doesn't. Just did a test poll to make sure and main text can be edited without affecting poll at all. It's only editing the response options that causes the reset.

**Comment by nehemiagurl** [comment]
huh. well, that sure doesn't seem consistent to me - I would presume that both kinds of editing can greatly affect the meaning of responses (see kjuh's example of a yes/no poll). 
I think changing it so editing the body also resets result is likely to upset some habits people have already made, so I believe notifying people of the edit and allowing them to change their vote would be better. it is more work to implement though, as it would also require implementing changing a vote.

**Comment by FediVideos** [comment]
Someone raised a good point on a Mastodon thread about this: even if changing voting was possible, edits that happen a short time before poll closes wouldn't leave enough time for people to change their vote, so the potential for abuse would still be there.

It would be a lot simpler just to give the poll creator a warning about a reset.

**Comment by gunchleoc** [comment]
Additionally, the text of the edit notification could be changed for participants in the poll to give them a hint that the poll was reset, so that they can consider voting again.

**Comment by richlv** [comment]
There seems to be another issue - if I vote on a poll, a typo gets fixed, my vote is removed - but I cannot vote again.
Also observed by others:
https://mastodon.social/@simonharper/111137918312160167

The current behaviour makes polls at least as unreliable as free editing would.

**Comment by nehemiagurl** [comment]
> There seems to be another issue - if I vote on a poll, a typo gets fixed, my vote is removed - but I cannot vote again.

Interesting, this doesn't seem to be a universal 4.2 thing. I re-voted on an edited poll just today, where the poll was both published and edited on 4.2, and I was voting from the same instance. Maybe something about the specific commit m.s is running.

**Comment by nettbeast** [comment]
If a poll or any part of it is edited, both the time and voters SHOULD reset.

If I create the following poll 

“Do you like Puppies?”
A. Yes
B. No

That’s fine, but if I go back and edit it to “Do you like Nazis” (In the header info) Votes should be reset, the timer should reset, and people should be notified so they can re-vote.

Any part of a poll changes it can fundamentally change the poll, whether it’s the asked question, or the poll options. And should force a reset.  (Now it should definitely WARN the poll’s poster that it’s happen, which is I think what the original post about it I saw was). But it needs to happen to prevent abuse.

**Comment by shaedrich** [comment]
> > Are you sure this isn't intentional? Being able to edit the context of a poll while retaining the results is not a good thing.
> 
> I would disagree. It makes the assumption that anyone posting has bad intentions just by posting a poll.
> 
> For example, if someone edits an option in the poll itself, then yes I would expect this to be the behavior. This would prevent from someone altering the results in some way based upon the options of the poll itself.
> 
> However, if someone makes a mistake in spelling of a word, the intent of the poll itself (the results) shouldn't suffer as a penalty.

Maybe, the question itself should have its own input field similar to poll options or CWs 🤔 

> Imagine a poll with „yes“ and „no“ as possible answers. A minor change to the polls text could easily turn the results upside down. While a possibility to fix typos is convenient, any possibilty to make changes conserving the answers opens an easy way to major manipulation.

Maybe, this can be improved with tying the edit history closer to the post:

![Image](https://github.com/user-attachments/assets/5ff86334-1cb4-4af8-9708-20b6c46e8d80)

(just a _very_ rough mockup)
